### ISSUE TITLE
back pressure on screen buffer

### ISSUE THREAD

**Issue by saurik** [mintty/mintty#1102]
In mintty--and notably, only mintty of any terminal I can remember using--if you accidentally cat a large binary file, it is effectively "game over" for your terminal: you can hit control-c all you want, but "the damage is done", as there will now be megabytes worth of extremely slow-to-paint block characters and likely-invalid terminal sequences that mintty will have buffered, and you will have to wait through all of it before you can reset the terminal and continue with your work (causing you to have to contemplate the tradeoff of losing your other state by instead just closing the whole terminal and opening a new one, starting over). The way this normally works is that there is minimal buffer over the network, causing an end-to-end flow control that prevents much more data having been generated than can be rendered locally; at the extreme, if you "stop" the local terminal, what normally happens is that this very quickly causes network buffers to fill up, preventing writes, which pushes back all the way to stdout on the appropriate tool. FWIW, I am not 100% sure if the issue is that the existing buffer is somewhat too large, if there is no buffer at all, or if the issue is "merely" that mintty is somehow literally hundreds of times slower than other terminals (causing even a small buffer to take forever to flush), but as I sit here typing this message, waiting for my terminal to recover, the end result feels similar ;P (oh! it just finished! I get to go back to work, lol).

**Comment by mintty** [comment]
I observed some instances of slow response when rendering some Unicode characters, esp. non-BMP, for the first time. The problem seemed to be on Windows side by my analysis. Whether that's your issue can only be judged with a sample of course.

**Comment by mintty** [comment]
Sounds reasonable and I'd like to look into this but I'd still want a sample so we have a reproducible case to analyse.

**Comment by saurik** [comment]
> if you accidentally cat a large binary file, it is effectively "game over" for your terminal: you can hit control-c all you want, but "the damage is done"

So: open a mintty Cygwin console, type "cat /bin/emacs" (emacs is popularly known to be a "large binary"), and then cry as control-C doesn't work correctly. FWIW, since Cygwin's emacs is actually "only" 5MB, you won't have to regret your decision _too_ long, but build yourself a larger file and the situation keeps getting worse.

**Comment by mintty** [comment]
As I don't have emacs, try `cat /bin/X` in xterm to compare. Not quite pleasant either. In general, catting binaries to a terminal is not a good idea. If you cat X in mintty and type ^C, it will stop after a few seconds, so what? I can look what to do about buffer size for a slightly faster response, but that has low priority.

**Comment by saurik** [comment]
I just did cat /bin/emacs in mintty and then immediately hit control-C. Using a stopwatch, I tracked how long it took for this command to complete and it took 70 seconds (a time which was easily reproduced when I tried again).

In contrast, I tried doing the same in Microsoft's "Bash on Ubuntu on Windows" terminal (I have no clue what this actually "is", terminal-pedigree wise): 1-2 seconds, at which point cat received an I/O error on write and terminated. I did the same thing using Terminal on macOS 10.14--where emacs is >20 times larger!--and it failed out (with no error printed, but the correct error code of 130) almost instantly.

I tried opening Windows Terminal, opening a Command Prompt instance, and then explicitly using Cygwin's cat.exe to cat Cygwin's emacs-nox.exe, and it reacts to Control-C within 1-2 seconds. Again, correctly, control-C is stopping the program, because if I don't hit control-C it happily will sit there for a long time printing out garbage... but, even after letting it spam, once I hit control-C, within 1-2 seconds it stops.

FWIW, mintty's behavior indicates that it does seem to eventually get around to causing the control-C: if I _don't_ hit control-C, it takes a very very long time to cat emacs (I know it takes at least three minutes... I haven't cared past that); the error code when you use control-C is also correct: 130. However, the fact that it it takes _70 seconds_ to react to control-C being pressed (which was consistent after waiting two minutes to press it) seems pretty obviously to be a defect very unique to this terminal.

I will also say that I figured out how to install X on Cygwin--mostly because I wanted to see how large the binary you were using was... it turned out to be half as large as the 5MB I'd noted for emacs--and it takes 25 seconds to react to control-C.

Notably, I tried "doubling up" /bin/X by first catting it twice into a single XX file and then catting that, and the behavior also only took 25 seconds, so it is feeling like the amount of time it is taking to react to control-C is in some way related to the amount of time it is taking to print or render characters in a buffer that is likely smaller than the size of the original X binary, and emacs simply has a higher ratio of slow-to-render characters.

On said note, I have now tried multiple fonts (I was using Lucida Console: I've also now tried Consolas and Courier New), turning off Font Smoothing (which I had previously set to "default"), and configuring a lower scrollback lines size (as I have a particularly large value set normally), and I can't make it faster than 25 seconds to cancel cat /bin/X.

The only thing I can come up with is that you fixed this very recently, as I'm using 3.4.7 (I'm in the middle of a project--one that happens to involve a lot of similarly-named files... some binary, some not--and I'm not in a good position to have to restart all my terminal sessions)... when I next have a chance, I'll try 3.5.0.

**Comment by BrianInglis** [comment]
Using `mintty` 3.4.7, Cygwin 3.2, Windows 10.0.19042.867 2020H2:
```
$ cat /proc/version
CYGWIN_NT-10.0-19042 version 3.2.0-340.x86_64 (corinna@calimero) (gcc version 9.3.0 20200312 (Fedora Cygwin 9.3.0-1) (GCC) ) 2021-03-29 08:42 UTC
$ time cat -A /bin/cygwin1.dll
...
real    0m6.093s
user    0m0.046s
sys     0m0.375s
$ time cat -A /bin/cygwin1.dll
...
^C
real    0m0.538s
user    0m0.000s
sys     0m0.061s```

**Comment by saurik** [comment]
@BrianInglis I mean, sure... if you are going to "cheat" the problem description by passing -A ;P. That same command takes me only 1.5s to run, but it takes almost _three minutes_ without -A. The problem is that by having a large buffer of slow-to-render terminal command sequences--which -A strips and sanitizes by replacing them all with ASCII simulacrums--you end up in a situation where a control-C right now can not have any visual effect on your terminal until a long time into the future (and even, depending on what is wrong with mintty where, situations where a control-C right now might not have an effect on the command itself until far into the future as well).

This also means you can't use "time" to measure it--though it is an interesting data point, as it helps us isolate whether the control-C is being held on the "sending" end or not, so this is actually an interesting test--because there might be an arbitrary amount of command sequences buffered before the output from time; and so, OK: I just did a cat of /bin/cygwin1.dll (without -A, of course) and immediately hit control-C; the command took, using a stopwatch, 40 seconds to complete... but time is claiming the cat command finished after only 20 seconds (and so the control-C "mattered" about halfway through execution).

One thing that does affect the execution time--as mintty's performance (which again: this isn't about; mintty should be allowed to render arbitrarily slowly and control-C should still work correctly) seems to be limited by "fill rate"--is making the terminal window smaller. If I leave the terminal window at its default size of 80x24 and do the "cat /bin/cygwin1.dll and immediately hit control-C" test it takes 15 seconds for the command to stop visually but ~7 seconds for time to report the command terminated (which is again about half-way through the total time; I feel like this might be elucidative to someone who worked on the terminal buffering).

**Comment by BrianInglis** [comment]
Assuming that mintty uses separate keyboard and/or mouse and/or pty input threads, and pty and/or window output threads, presumably ctrl-{O,C,\} is received immediately as a message and sent immediately to the pty, where the console code threads forward it to the shell, which runs commands, and returns output to console code, to return to the pty? 

One of my first questions about having the console code handle any escape sequences was where do you draw that line, as I know from days of yore, that many devices had many escape sequences where you sent large chunks of data in various encodings as well as raw binary? 
Most of those devices and I/O "drivers" used DFAs with timeouts for bad sequences, and shortcuts for speedups and efficiency handling common cases like colours. 

Might be time for the OP to post on the Cygwin list to see what the console developer has to say about handling large random binary outputs.

**Comment by mintty** [comment]
@BrianInglis, mintty control is basically event-driven, handling Windows keyboard/mouse events (in common) and checking for new pty input, not using threads. Escape sequences, once collected, are handled immediately, i.e. sequentially, not scheduled in any way. Handling of bad sequences might be a point to consider. I can hardly imagine any shortcuts in handling control sequences as that would usually violate the terminal interaction model. Particularly I don't see a reliable way to apply a shortcut for ^C as that is just a normal character from the terminal point of view.

**Comment by mintty** [comment]
Replacing some controls characters in a binary (cygwin1.dll) suggests that BEL (^G) may be the main culprit.
There is some acceleration for BEL already, but if you confirm this for your samples, I can look into BEL handling again.

**Comment by avih** [comment]
> suggests that BEL (^G) may be the main culprit.

It's definitely _a_ culprit, and the speed does seem to improve greatly when filtering it out.

E.g.:
```sh
$ time cat `which man`  # just less than 100K
...
real    0m0.097s
user    0m0.031s
sys     0m0.046s
```
So it completes quickly enough as far as `time` is concerned, but with an external stopwatch it took 10 seconds until the prompt appeared.

However, running this:
```sh
$ time cat `which man` | tr "$(printf \\007)" "?"  # replace \g with ?
...
real    0m0.098s
user    0m0.015s
sys     0m0.030s
```
I.e. pretty much the same duration as far as `time` is concerned, but with an external stopwatch the prompt appears after 3 seconds.

So yeah, 3x faster is definitely a meaningful improvement, but still, 3s for 100K sounds a bit much.

For reference, on the same system, in `xterm` the first command completes (with an external stopwatch) after about 1s, while the second command (filtered output) is quite faster - I'm estimating 200-300 ms.

In `st` on the same system, the same pair of commands complete in roughly the same speed - faster than in `xterm`, and I'm estimating about 100 ms or less.

**Comment by avih** [comment]
> ```shell
> $ time cat `which man` | tr "$(printf \\007)" "?"  # replace \g with ?
> ```

For reference, there are 179 bytes of ^G at this file:
```sh
$ cat `which man` | od -bvAn | tr " " "\n" | grep -c 007
179
```

**Comment by mintty** [comment]
There was a screen update flush on every BEL character, even if the sound was squashed within a BEL series.
Please test the tweak.

**Comment by avih** [comment]
> Please test the tweak.

I don't have a mintty build setup ready.

I assume you did 1. reproduce numbers similar to mine with and without filtering out ^G, and 2. confirm that this tweak brings the numbers closer/down-to the filtered numbers?

So this test is only to confirm the improvement you already measured?

**Comment by BrianInglis** [comment]
> So it completes quickly enough as far as `time` is concerned, but with an external stopwatch it took 10 seconds until the prompt appeared.

> I.e. pretty much the same duration as far as `time` is concerned, but with an external stopwatch the prompt appears after 3 seconds.
> 
> So yeah, 3x faster is definitely a meaningful improvement, but still, 3s for 100K sounds a bit much.
> 
> For reference, on the same system, in `xterm` the first command completes (with an external stopwatch) after about 1s, while the second command (filtered output) is quite faster - I'm estimating 200-300 ms.
> 
> In `st` on the same system, the same pair of commands complete in roughly the same speed - faster than in `xterm`, and I'm estimating about 100 ms or less.

Instead of a stopwatch, you could bracket the timed test with commands: 
```
$ date +%s.%N
1620583824.138215000
```

**Comment by avih** [comment]
> Instead of a stopwatch, you could bracket the timed test with commands:

I don't think I understand. I'd guess that if you print the date before and after the run then it would still produce results same as with `time`.

So could you please write the full command which will print a result similar to what an external stopwatch measures?

**Comment by mintty** [comment]
> I don't have a mintty build setup ready.

Quite basic. Just gcc and make packages are needed.

**Comment by saurik** [comment]
(FWIW, I agree that compiling mintty is very easy. There is a ridiculous issue I had while compiling it today, but the bug doesn't seem to be in mintty's build scripts... it almost seems like a bug in windres, but it seems too basic for that, so I'm wondering if I broke it somehow. I worked around it quickly.)

I have compiled a new build of mintty, and used a "local" stopwatch (which is the correct way to test this: a "remote" time or date pair is going to measure the wrong latency... that said, currently, these measurements _do_ happen to be related, which I think should be elucidative of something) to test.

The time to cancel a cat of /bin/emacs is down from ~70s to ~7s, which is a massive improvement. Thank you!! That said, it still isn't clear to me that the "root cause" is being addressed; but, after sitting here for a half hour carefully reading src/client.c, I'm thinking the issue might be in Cygwin's pty.

**Comment by mintty** [comment]
> a ridiculous issue I had while compiling it today

Yeah, I had forgotten about this nuisance, sorry. A regression bug was introduced in binutils 2.36. Workaround for now: downgrade to binutils 2.35 and save a copy of /bin/windres into your local path before you upgrade again.

**Comment by saurik** [comment]
My workaround was much simpler (and way more hilarious ;P).

```
$ cat /bin/'gcc -E -xc -DRC_INVOKED -MMD -MP -DTARGET=x86_64-pc-cygwin -DNDEBUG'
#!/bin/bash
exec $0 "$@"
```

**Comment by avih** [comment]
> Quite basic. Just gcc and make packages are needed.

It was a euphemism for "I don't quite feel like building it right now".

**Comment by mintty** [comment]
Released 3.5.1.
